lua/ibl/autocmds.lua

```diff
@@ -0,0 +1,61 @@
+local highlights = require "ibl.highlights"
+local M = {}
+
+M.setup = function()
+    local group = vim.api.nvim_create_augroup("IndentBlankline", {})
+    local ibl = require "ibl"
+    local buffer_leftcol = {}
+
+    vim.api.nvim_create_autocmd("VimEnter", {
+        group = group,
+        pattern = "*",
+        callback = ibl.refresh_all,
+    })
+    vim.api.nvim_create_autocmd({
+        "CursorMoved",
+        "BufWinEnter",
+        "CompleteChanged",
+        "FileChangedShellPost",
+        "FileType",
+        "TextChanged",
+        "TextChangedI",
+    }, {
+        group = group,
+        pattern = "*",
+        callback = function(opts)
+            ibl.debounced_refresh(opts.buf)
+        end,
+    })
+    vim.api.nvim_create_autocmd("OptionSet", {
+        group = group,
+        pattern = "list,listchars,shiftwidth,tabstop,vartabstop",
+        callback = function(opts)
+            ibl.debounced_refresh(opts.buf)
+        end,
+    })
+    vim.api.nvim_create_autocmd("WinScrolled", {
+        group = group,
+        pattern = "*",
+        callback = function(opts)
+            local win_view = vim.fn.winsaveview() or { leftcol = 0 }
+
+            if buffer_leftcol[opts.buf] ~= win_view.leftcol then
+                buffer_leftcol[opts.buf] = win_view.leftcol
+                -- Refresh immediately for horizontal scrolling
+                ibl.refresh(opts.buf)
+            else
+                ibl.debounced_refresh(opts.buf)
+            end
+        end,
+    })
+    vim.api.nvim_create_autocmd("ColorScheme", {
+        group = group,
+        pattern = "*",
+        callback = function()
+            highlights.setup()
+            ibl.refresh_all()
+        end,
+    })
```

Comment:
Isn't this a bit strict?

I cannot for example use a custom hlg, which is not set by the colorscheme itself.

`minimal.lua`:
```lua
-- DO NOT change the paths and don't remove the colorscheme
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath, })
end
vim.opt.runtimepath:prepend(lazypath)

vim.cmd [[hi MyHLG guifg=red]]

-- install plugins
local plugins = {
    {
        "lukas-reineke/indent-blankline.nvim",
        branch = "v3",
        config = function()
            require("ibl").setup {
                indent = { highlight = "MyHLG" },
            }
        end
    },
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

vim.cmd [[set termguicolors]]
```

Now when I run for example `colorscheme default`, it clears `MyHLG`, and when the autocommand runs it errors

Reply:
You need to use the `HIGHLIGHT_SETUP` hook, it's explained in the readme and the docs

https://github.com/lukas-reineke/indent-blankline.nvim/tree/v3#multiple-indent-colorsh

For the visual selection, I can't reproduce that. Please open an issue with steps to get there.

Reply:
Hmm, yes you are right. I just read `:h ibl.setup()` and  scrolled to `config.indent`. Out of curiosity — is there a reason why hooks are necessary here? None of other plugins I use with custom hl groups require that (`tzachar/highlight-undo.nvim`, `winston0410/range-highlight.nvim`, `folke/flash.nvim` to name a few), and their config is more concise. 

About the highlighting — I'm not sure it is an "issue" per see, but side effect of how highlights are handled as well. Try this

```lua
local root = vim.fn.fnamemodify("./.repro", ":p")
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath, })
end
vim.opt.runtimepath:prepend(lazypath)

local plugins = {
    "RRethy/nvim-base16",
    {
        "lukas-reineke/indent-blankline.nvim",
        branch = "v3",
        config = function()
            require("ibl").setup { }
        end
    },
}

require("lazy").setup(plugins, {
    root = root .. "/plugins",
})

vim.cmd [[set termguicolors]]
vim.cmd [[colorscheme base16-dirtysea]]
vim.opt.background = "light"
vim.api.nvim_set_hl(0, "NonText",  { fg = "Gray", bg = "bg" })
```

Anyway, thanks for the hard work on this plugin!

Reply:
I agree it's not ideal, but I don't see a better way. Because there can be different colors for the indent guides, the scope, and the whitespace, I need to generate all combinations for all cases where they are used.

Otherwise, to have 3 colors each, the user would need to manually make 42 highlight groups. That would be much less user-friendly.
Even with just one color each, I still need to create 7 highlight groups.